File: android/app/src/main/java/ae/findmybay/feature/home/NearbyViewModel.kt
package ae.findmybay.feature.home

import android.annotation.SuppressLint
import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import ae.findmybay.data.repo.VendorRepository
import ae.findmybay.domain.model.Vendor
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.google.android.gms.tasks.CancellationTokenSource
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

data class NearbyUiState(
    val loading: Boolean = false,
    val lat: Double? = null,
    val lng: Double? = null,
    val vendors: List<Vendor> = emptyList(),
    val error: String? = null,
    val needsPermission: Boolean = true,
    // Filter state — applied client-side over [vendors] to produce the
    // [filteredVendors] view used by the map markers and peek card.
    val searchQuery: String = "",
    val freeNowOnly: Boolean = false,
    val under40Only: Boolean = false,
) {
    /**
     * The list after applying search + filter chips. Computed eagerly here
     * so the screen doesn't have to re-derive on every recomposition.
     */
    val filteredVendors: List<Vendor>
        get() {
            val q = searchQuery.trim().lowercase()
            return vendors.filter { v ->
                val matchesQuery = q.isEmpty() ||
                    v.brandName.lowercase().contains(q) ||
                    v.city.lowercase().contains(q)
                val matchesFree = !freeNowOnly || v.freeBays > 0
                val matchesPrice = !under40Only || (v.priceFromAed != null && v.priceFromAed < 40)
                matchesQuery && matchesFree && matchesPrice
            }
        }
}

@HiltViewModel
class NearbyViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
    private val vendors: VendorRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(NearbyUiState())
    val state: StateFlow<NearbyUiState> = _state.asStateFlow()

    private val fused by lazy { LocationServices.getFusedLocationProviderClient(context) }

    fun onPermissionGranted() {
        _state.update { it.copy(needsPermission = false) }
        refresh()
    }

    fun onPermissionDenied() {
        // Fall back: center map on Dubai's Burj Khalifa with no auto-load.
        _state.update {
            it.copy(
                needsPermission = true,
                lat = 25.1972,
                lng = 55.2744,
                error = "Allow location access to see car washes near you.",
            )
        }
    }

    @SuppressLint("MissingPermission") // caller has already granted at this point
    fun refresh() {
        _state.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            val location = runCatching {
                val cts = CancellationTokenSource()
                fused.getCurrentLocation(Priority.PRIORITY_BALANCED_POWER_ACCURACY, cts.token).await()
            }.getOrNull()

            if (location == null) {
                _state.update {
                    it.copy(
                        loading = false,
                        error = "Couldn't read your location. Try again outdoors or with GPS on.",
                    )
                }
                return@launch
            }

            _state.update { it.copy(lat = location.latitude, lng = location.longitude) }

            val list = runCatching { vendors.nearby(location.latitude, location.longitude) }
                .getOrElse {
                    _state.update { s -> s.copy(loading = false, error = friendly(it)) }
                    return@launch
                }

            _state.update { it.copy(loading = false, vendors = list) }
        }
    }

    // ── Filter setters ──────────────────────────────────────────────────

    fun setSearchQuery(q: String) = _state.update { it.copy(searchQuery = q) }
    fun toggleFreeNow() = _state.update { it.copy(freeNowOnly = !it.freeNowOnly) }
    fun toggleUnder40() = _state.update { it.copy(under40Only = !it.under40Only) }

    private fun friendly(e: Throwable): String =
        e.message?.takeIf { it.isNotBlank() } ?: "Couldn't load car washes nearby."
}

File: ios/FindMyBay/Features/Home/NearbyViewModel.swift
// NearbyViewModel.swift
//
// State for the home/map screen — location, vendor list, search, filters.
// Mirrors `feature/home/NearbyViewModel.kt`.

import Foundation
import Observation
import CoreLocation

@Observable @MainActor
final class NearbyViewModel {

    // MARK: - State (matches NearbyUiState)
    var loading: Bool = false
    var lat: Double? = nil
    var lng: Double? = nil
    var vendors: [Vendor] = []
    var error: String? = nil
    var needsPermission: Bool = true

    // Filter chips (computed against `vendors` to produce `filteredVendors`)
    var searchQuery: String = ""
    var freeNowOnly: Bool = false
    var under40Only: Bool = false

    /// View used by the markers + peek card. Eagerly computed via getter to
    /// match the Kotlin `filteredVendors` property.
    var filteredVendors: [Vendor] {
        let q = searchQuery.trimmingCharacters(in: .whitespaces).lowercased()
        return vendors.filter { v in
            let matchesQuery = q.isEmpty
                || v.brandName.lowercased().contains(q)
                || v.city.lowercased().contains(q)
            let matchesFree  = !freeNowOnly || v.freeBays > 0
            let matchesPrice = !under40Only || (v.priceFromAed.map { $0 < 40 } ?? false)
            return matchesQuery && matchesFree && matchesPrice
        }
    }

    // MARK: - Deps
    private let repo: VendorRepository
    private let location: LocationService

    init(repo: VendorRepository, location: LocationService = .shared) {
        self.repo = repo
        self.location = location
    }

    // MARK: - Permissions

    func bootstrap() {
        Task {
            let granted = await location.requestPermission()
            if granted {
                onPermissionGranted()
            } else {
                onPermissionDenied()
            }
        }
    }

    func onPermissionGranted() {
        needsPermission = false
        refresh()
    }

    func onPermissionDenied() {
        // Match the Android fallback — center on Burj Khalifa, no auto-load.
        needsPermission = true
        lat = 25.1972
        lng = 55.2744
        error = LocationError.permissionDenied.errorDescription
    }

    // MARK: - Refresh

    func refresh() {
        loading = true
        error = nil
        Task {
            do {
                let location = try await location.currentLocation()
                lat = location.coordinate.latitude
                lng = location.coordinate.longitude
                let list = try await repo.nearby(lat: location.coordinate.latitude,
                                                 lng: location.coordinate.longitude)
                vendors = list
                loading = false
            } catch {
                loading = false
                self.error = (error as? LocalizedError)?.errorDescription
                    ?? "Couldn't load car washes nearby."
            }
        }
    }

    // MARK: - Filter setters

    func setSearchQuery(_ q: String) { searchQuery = q }
    func toggleFreeNow()             { freeNowOnly = !freeNowOnly }
    func toggleUnder40()             { under40Only = !under40Only }
}
